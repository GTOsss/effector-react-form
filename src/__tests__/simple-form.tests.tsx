import React from 'react';
import {createStore} from 'effector';
import {render, fireEvent, screen} from '@testing-library/react';
import {useForm} from '../';
import {Controller, OnSubmit} from '../../index';

interface Values {
  username?: string,
  profile?: {
    firstName?: string,
    lastName?: string,
  }
}

interface InputProps {
  controller: Controller,
  label: string,
}

const renderForm = () => {
  const $fieldsInline = createStore({});

  const Input: React.FC<InputProps> = ({
    controller,
    label,
  }) => {
    const {input, fieldState, form, error} = controller();

    const isShowError = error && (form.submitted || form.forcedError || fieldState.blurred);

    return (
      <div role="wrapper-for-input" className="input-wrap">
        <label>{label}</label>
        <input
          {...input}
          value={input.value || ''}
          role="textbox"
          className="input"
          placeholder={label}
        />
        {isShowError && (<span>{error}</span>)}
      </div>
    );
  };

  const SimpleForm = () => {
    const {handleSubmit, controller, setOrDeleteError} = useForm({$fieldsInline});

    const onSubmit: OnSubmit<Values> = () => {
    };

    return (
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          label="Username"
          controller={controller({name: 'username'})}
        />
        <Input
          label="First name"
          controller={controller({name: 'profile.firstName'})}
        />
        <Input
          label="Last name"
          controller={controller({name: 'profile.lastName'})}
        />
        <button type="submit">submit</button>
        <button
          type="button"
          onClick={() => setOrDeleteError({
            field: 'profile.firstName',
            error: 'First name is not valid',
          })}
        >
          set error for firstName
        </button>
      </form>
    );
  };

  render(<SimpleForm />);

  return {$fieldsInline};
};

describe('SimpleForm', () => {

  test('$fieldsInline after render', () => {
    const {$fieldsInline} = renderForm();
    expect($fieldsInline.getState()).toMatchSnapshot();
  });

  test('onChange username', () => {
    renderForm();
    const input = screen.getByPlaceholderText('Username');
    fireEvent.change(input, {target: {value: 'test'}});
    const inputs = screen.getAllByRole('textbox');
    expect(inputs).toMatchSnapshot();
  });

  test('onChange profile.fistName', () => {
    renderForm();
    const input = screen.getByPlaceholderText('First name');
    fireEvent.change(input, {target: {value: 'test'}});
    const inputs = screen.getAllByRole('textbox');
    expect(inputs).toMatchSnapshot();
  });

  test('setError profile.firstName', () => {
    renderForm();
    const button = screen.getByText('set error for firstName');
    fireEvent.click(button);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });

  test('setError profile.firstName and submit', () => {
    renderForm();
    const button = screen.getByText('set error for firstName');
    fireEvent.click(button);
    const buttonSubmit = screen.getByText('submit');
    fireEvent.click(buttonSubmit);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  });
});


