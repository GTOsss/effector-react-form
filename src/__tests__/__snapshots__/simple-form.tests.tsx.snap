// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleForm $fieldsInline after render 1`] = `
Object {
  "profile.firstName": Object {
    "active": false,
    "blurred": false,
    "changed": false,
    "touched": false,
  },
  "profile.lastName": Object {
    "active": false,
    "blurred": false,
    "changed": false,
    "touched": false,
  },
  "username": Object {
    "active": false,
    "blurred": false,
    "changed": false,
    "touched": false,
  },
}
`;

exports[`SimpleForm onChange profile.fistName 1`] = `
Array [
  <input
    class="input"
    name="username"
    placeholder="Username"
    role="textbox"
    value=""
  />,
  <input
    class="input"
    name="profile.firstName"
    placeholder="First name"
    role="textbox"
    value="test"
  />,
  <input
    class="input"
    name="profile.lastName"
    placeholder="Last name"
    role="textbox"
    value=""
  />,
]
`;

exports[`SimpleForm onChange username 1`] = `
Array [
  <input
    class="input"
    name="username"
    placeholder="Username"
    role="textbox"
    value="test"
  />,
  <input
    class="input"
    name="profile.firstName"
    placeholder="First name"
    role="textbox"
    value=""
  />,
  <input
    class="input"
    name="profile.lastName"
    placeholder="Last name"
    role="textbox"
    value=""
  />,
]
`;

exports[`SimpleForm setError profile.firstName 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      name="username"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      First name
    </label>
    <input
      class="input"
      name="profile.firstName"
      placeholder="First name"
      role="textbox"
      value=""
    />
    <span>
      First name is not valid
    </span>
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Last name
    </label>
    <input
      class="input"
      name="profile.lastName"
      placeholder="Last name"
      role="textbox"
      value=""
    />
  </div>,
]
`;

exports[`SimpleForm setError profile.firstName and submit 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      name="username"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      First name
    </label>
    <input
      class="input"
      name="profile.firstName"
      placeholder="First name"
      role="textbox"
      value=""
    />
  </div>,
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Last name
    </label>
    <input
      class="input"
      name="profile.lastName"
      placeholder="Last name"
      role="textbox"
      value=""
    />
  </div>,
]
`;
